File: tool/feedback/src/main/java/feedback/time/TimeDifference.java
package feedback.time;

import feedback.parser.DistributedLog;
import org.joda.time.DateTime;

public final class TimeDifference {
    public final DateTime good, bad;
    public final long difference;  // bad - good

    public TimeDifference(final DateTime good, final DateTime bad) {
        this.good = good;
        this.bad = bad;
        this.difference = this.bad.getMillis() - this.good.getMillis();
    }

    public TimeDifference(final DistributedLog goodRun, final DistributedLog badRun) {
        DateTime good = goodRun.logs[0].entries[0].datetime, bad = badRun.logs[0].entries[0].datetime;
        for (int i = 1; i < goodRun.logs.length; i++) {
            if (good.isBefore(goodRun.logs[i].entries[0].datetime)) {
                good = goodRun.logs[i].entries[0].datetime;
                bad = badRun.logs[i].entries[0].datetime;
            }
        }
        this.good = good;
        this.bad = bad;
        this.difference = this.bad.getMillis() - this.good.getMillis();
    }

    public TimeDifference(final DateTime good, final DateTime bad, final TimeDifference timeDifference) {
        this(timeDifference.good2bad(good), bad);
    }

    // a valid difference must < 1 hour = 3,600,000 ms
    public int getTimeScore() {
        if (difference < 0) {
            return -3 * (int) difference;
        }
        return (int) difference;
    }

    public DateTime good2bad(final DateTime good) {
        return good.plus(this.difference);
    }
}

File: tool/feedback/src/main/scala/feedback/time/TimeDifference.scala
package feedback.time

import feedback.log.{DistributedWorkloadLog, Log, LogFile, UnitTestLog}
import org.joda.time.DateTime

final class TimeDifference(val good: DateTime, val bad: DateTime) {
  val difference: Long = bad.getMillis - good.getMillis

  def this(pair: (DateTime, DateTime)) = this(pair._1, pair._2)
  def this(good: LogFile, bad: LogFile) = this(good.showtime, bad.showtime)

  def this(good: Log, bad: Log) = this((good, bad) match {
    case (UnitTestLog(good, _), UnitTestLog(bad, _)) =>
      (good.showtime, bad.showtime)
    case (DistributedWorkloadLog(good), DistributedWorkloadLog(bad)) =>
      require(good.length == bad.length)
      good.map(_.showtime).zip(bad.map(_.showtime)).min
  })

  def good2bad(good: DateTime): DateTime = good.plus(difference)
}
